pandas/core/reshape/reshape.py

```diff
@@ -643,8 +643,15 @@ def _convert_level_number(level_num, columns):
             levs.append(np.take(lev, level_codes))
         tuples = list(zip(*levs))
         unique_groups = [key for key, _ in itertools.groupby(tuples)]
-        new_names = this.columns.names[:-1]
-        new_columns = MultiIndex.from_tuples(unique_groups, names=new_names)
+        new_columns = MultiIndex.from_arrays(
+            [
+                Index(new_level, dtype=level.dtype)
+                if None not in new_level
+                else new_level
+                for new_level, level in zip(zip(*unique_groups), this.columns.levels)
+            ],
```

Comment:
pandas/core/reshape/reshape.py:657: error: Incompatible types in assignment (expression has type "MultiIndex", variable has type "List[Any]")  [assignment]

is coming up

umm prob need a cast there

Reply:
Done.